path_pattern/pattern.go

```diff
@@ -23,33 +24,45 @@ func (p *Pattern) UnmarshalText(data []byte) error {
 	return nil
 }
 
-// Match happens if the prefix of the input matches the pattern.
+func (p Pattern) regexp() *regexp.Regexp {
+	var pieces []string
+	for _, piece := range p {
+		pieces = append(pieces, piece.Regexp())
+	}
+	return regexp.MustCompile("^" + strings.Join(pieces, "/") + "$")
+}
+
+// Match happens if the pattern exactly matches the string.
 func (p Pattern) Match(v string) bool {
-	parts := strings.Split(v, "/")
+	r := p.regexp()
+	return r.MatchString(removeTrailingSlashes(v))
+}
 
-	compLen := len(p)
-	if compLen > len(parts) {
-		return false
-	}
+func (p Pattern) MatchWithGroup(v string) (bool, []string) {
+	r := p.regexp()
+	subMatches := r.FindStringSubmatch(removeTrailingSlashes(v))
+	return subMatches != nil, subMatches
+}
 
-	for i := 0; i < compLen; i++ {
-		if !p[i].Match(parts[i]) {
-			return false
-		}
+// Removes trailing slashes, except the final slash if v == "/".
+func removeTrailingSlashes(v string) string {
+	s := v
+	for len(s) > 1 && s[len(s) - 1] == '/' {
+		s = s[:len(s) - 1]
 	}
-	return true
+	return s
 }
 
 // Converts a string pattern "/v1/{arg2}" to Pattern.
 func Parse(v string) Pattern {
-	parts := strings.Split(v, "/")
+	parts := strings.Split(removeTrailingSlashes(v), "/")
 	result := make(Pattern, 0, len(parts))
```

Comment:
I am not really clear on how trailing slashes are handled.  From the test it looks like
/foo/{arg2}
matches
/foo/a/
which is fine, but maybe a comment somewhere should set expectations.
?

Reply:
It matches ignoring trailing slashes, so `/v1/{arg}` and `/v1/{arg}/` would both match `/v1/foo` and `/v1/foo/`.  I added a comment.